Parameterize Neo4j version in container tests

fbiville opened the pull request.

fbiville committed:
Parameterize Neo4j version in container tests

fbiville commented on .github/workflows/main.yml:
```diff
@@ -9,12 +9,24 @@ on:
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        java: [ 8, 11, 16 ]
```
shall I test only Java 8?
Only 8 and 11?

fbiville replied:
I'll reduce to Java 8 for now, since that would otherwise generate a LOT of builds

fbiville commented on neo4j-jdbc-bolt/src/test/java/org/neo4j/TestcontainersCausalCluster.java:
```diff
@@ -125,7 +125,7 @@ private static Neo4jContainer getNeo4jContainer(WaitStrategy waitForBolt,
                                                                       Map<String, Object> neo4jConfig,
                                                                       String name) {
-        Neo4jContainer container = (Neo4jContainer) new Neo4jContainer("neo4j:3.5-enterprise")
+        Neo4jContainer container = (Neo4jContainer) new Neo4jContainer("neo4j:3.5-enterprise") // TODO: parameterize this?
```
not sure if this one should be parameterized or not

conker84 replied:
I think that we can use the same approach that we have here:

https://github.com/neo4j-contrib/neo4j-jdbc/blob/fae9eeb895755a703d814def94c2ddd5a0870fd8/neo4j-jdbc-bolt/src/test/java/org/neo4j/jdbc/bolt/BoltNeo4jAuthenticationIT.java#L49

fbiville replied:
I just hope that the cluster configuration settings have not changed since 3.5 😨

conker84 replied:
I mean by using the method `neo4jImageCoordinates()`

conker84 replied:
> I just hope that the cluster configuration settings have not changed since 3.5 😨

Maybe we're lucky!

fbiville replied:
Turns out this class is used only by a single, ignored, test class: `BoltRoutingIT`. I migrated it anyway.

fbiville committed:
Work around multi-tenancy differences across major and minor versions

conker84 reviewed (approved):
Just the small comment about the cluster node docker image, otherwise LGTM

fbiville committed:
Fix DatabaseMetaData#getIndexInfo implementation for Neo4j 3.5

Neo4j 3.5 db.indexes proc does not yield a label column and the
index name and qualifier is slightly different

fbiville committed:
Avoid sharing state in ResultSet integration tests

Make the container rule non-static. This is slower but makes the
tests easier to reason about.

fbiville committed:
Fix GH actions branch declaration to comply to schema

fbiville committed:
Parameterize container in cluster creation utility

Fix warnings in the process

conker84 committed:
clear cache state when we close connection in order to prevent flaky tests

conker84 committed:
fix mocking problem Co-authored-by: fbiville <florent.biville@neo4j.com>

conker84 committed:
added ordering  Co-authored-by: fbiville <florent.biville@neo4j.com>

conker84 committed:
improved cache clear into the tests Co-authored-by: fbiville <florent.biville@neo4j.com>

conker84 committed:
fixed flaky test Co-authored-by: fbiville <florent.biville@neo4j.com>

conker84 committed:
moved cache cleanup from Before to After actions Co-authored-by: fbiville <florent.biville@neo4j.com>
